asmcomp/cmm.ml

```diff
@@ -222,3 +222,51 @@ let ccatch (i, ids, e1, e2, dbg) =
 
 let reset () =
   label_counter := 99
+
+let lift_phantom_lets f exp =
+  (* Unbounded exploration would be linear in the size of the expression,
+     which can lead to a translation pass that becomes quadratic in the size
+     of the original program. Limiting to a fixed depth prevents this problem,
+     and it happens that no existing optimisation looks deeper than 4 nested
+     operations, so no optimisation is lost because of this restriction. *)
+  let rec lift exp lifted_rev depth =
+    let next_depth = depth + 1 in
+    match exp with
+    | Cop (op, args, dbg) when depth < 4 ->
+      let lifted_rev, args =
+        List.fold_left (fun (lifted_rev, args) arg ->
+            let lifted_rev, arg = lift arg lifted_rev next_depth in
+            lifted_rev, arg :: args)
+          ([], [])
+          (List.rev args)
+      in
+      lifted_rev, Cop (op, args, dbg)
+    | Cphantom_let (var, defining_expr, body) ->
+      lift body ((var, defining_expr) :: lifted_rev) depth
+    | _ -> lifted_rev, exp
+  in
+  let lifted_rev, exp = lift exp [] 0 in
+  let exp = f exp in
+  List.fold_left (fun exp (var, defining_expr) ->
+      Cphantom_let (var, defining_expr, exp))
+    exp
+    lifted_rev
+
+
+let rec map_single_tail f = function
+  | Clet (id, exp, body) ->
+      let result = map_single_tail f body in
+      Clet (id, exp, result)
+  | Cphantom_let (var, defining_expr, body) ->
```

Comment:
@lthls : I guess we should remove the phantom let case here, right? (I forgot to ask this for your PR)

Reply:
No, in cases like `Cphantom_let (_, _, Clet (_, _, body))`, if you remove the `Cphantom_let` case from `map_single_tail` then `lift_phantom_lets` would not traverse the `Clet` construction, and the function being mapped would get `Clet (_, _, body)` instead of `body` as argument.

For the record, the rationale for keeping two different functions is that `lift_phantom_lets` needs to be able to remove deep phantom lets, but since phantom lets are not translated to code it's safe to move them around without thinking too much about it (though putting them in the wrong order can lead to unusable debug information). On the other and, `map_single_tail` needs to look through code-carrying constructions, so it makes sense to keep it simple enough that assessing its correctness is straightforward.